src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,301 @@
+                        // or if one of its dependencies is dirty
+                        bool isDirty = info.IsDirty;
+                        foreach (var dependency in info.UsedLocalFunctions)
+                        {
+                            isDirty |= LocalFuncResults[dependency].IsDirty;
+                            if (isDirty)
+                            {
+                                break;
+                            }
+                        }
+
+                        info.IsDirty = false;
+                        if (isDirty)
+                        {
+                            // Note: Recursively visits this local function, which
+                            // may cause other local functions to be visited, even
+                            // if they're not marked dirty.
+                            // Consider: Is there a more efficient walk that we could
+                            // do here?
+                            VisitLambdaOrLocalFunction(info.BoundNode);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static bool HasDirtyLocalFunctions<T>(T localFuncInfos)
+            where T : IEnumerable<LocalFuncInfo>
+        {
+            foreach (var info in localFuncInfos)
+            {
+                if (info.IsDirty)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private void NoteReadInLocalFunction(Symbol variable, Symbol capturedSymbol, int? slotOpt = null)
+        {
+            var localFunc = GetNearestLocalFunctionOpt();
+
+            Debug.Assert(localFunc != null);
+
+            var results = LocalFuncResults[localFunc];
+            var usedVar = new LocalFuncInfo.UsedVariable(written: false,
+                                                         capturedSymbol: capturedSymbol,
+                                                         symbol: variable,
+                                                         slot: slotOpt);
+            if (!results.UsedVariables.Contains(usedVar))
+            {
+                results.UsedVariables.Add(usedVar);
+                results.IsDirty = true;
+            }
+        }
+
+        private void AssignVarInLocalFunction(int slot, Symbol capturedSymbol)
+        {
+            var localFunc = GetNearestLocalFunctionOpt();
+
+            Debug.Assert(localFunc != null);
+
+            var info = LocalFuncResults[localFunc];
+            var usedVar = new LocalFuncInfo.UsedVariable(written: true,
+                                                         capturedSymbol: capturedSymbol,
+                                                         slot: slot);
+            if (!info.UsedVariables.Contains(usedVar))
+            {
+                info.UsedVariables.Add(usedVar);
+                info.IsDirty = true;
+            }
+        }
```

Comment:
I see that this records that the variable was written. But we need to know if the variable was written *on every execution path through the local function*.